src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/util/InstanceDetails.java

```diff
@@ -15,48 +15,166 @@
 
 package com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.util;
 
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.core.Util;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.reader.ClusterDetailsEventProcessor;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+import com.google.common.net.InetAddresses;
 
 public class InstanceDetails {
+  public static class Ip {
+
+    // The only way to get the ip is to get the serialized string representation of it.
+    private String ip;
+
+    public Ip(String ip) {
+      if (!InetAddresses.isInetAddress(ip)) {
+        throw new IllegalArgumentException("The provided string is not an IPV4ip: '" + ip + "'");
+      }
+      this.ip = ip;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof Ip)) {
+        return false;
+      }
+      Ip ip1 = (Ip) o;
+      return Objects.equal(ip, ip1.ip);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(ip);
+    }
+
+    @Override
+    public String toString() {
+      return "" + ip;
```

Comment:
I'm curious, what does this do differently than just directly returning ip?

Reply:
you mean why we don't have `getIp()` instead ?

Reply:
I mean why `return "" + ip` instead of just `return ip`?

Reply:
Yes you are right. It's a no-op for a string